### PR TITLE
microdds_client: support Agent IP address as PX4 parameter

### LINKED ISSUE #21179: microdds client: add Agent IP as PX4 parameter
## Describe problem solved by the proposed feature
Differently from _mavlink_, when UDP transport protocol is used, the _microdds client_ requires the explicit IP address of the companion computer running the  _agent_.
This value is currently passed to the module as an argument at startup in the form
```sh
microdds_client start -t udp -p 8888 -h 192.168.0.1
```
which tells that UPD communication to an Agent running at `192.168.0.1` and listening to the port `8888` should be used.
If not given, the default IP value is the loopback address `127.0.0.1` which is just right for simulations.
However, when working on real drones this prevents starting the client from PX4 parameters ( `XRCE_DDS_0_CFG=ethernet`) as there is no PX4 parameter for storing the Agent IP address. As a consequence, the default value is always used.

https://github.com/PX4/PX4-Autopilot/blob/2008a447c352fb04d9b817ec01593c66a92d7bae/src/modules/microdds_client/module.yaml#L14-L33

## Describe your preferred solution
It would be great to have the Agent IP address as a PX4 parameter. One INT32 parameter encoding the address would be enough and the client would just have to parse it.

## Describe possible alternatives

- Instead of using 1 parameter, a more user friendly solution could be to use 4 separate parameters to encode the address as PARAM1:PARAM2:PARAM3:PARAM4.
- Implement _string_ parameters?
- Remove the `supports_networking: true` from `XRCE_DDS_CFG` and force the users to use [system startup](https://docs.px4.io/main/en/concept/system_startup.html) if they want to use UDP transport protocol.

### PR DESCRIPTION
**Pull request by beniaminopozzan**
### Features Implemented

1. Removed incomplete multi instance capability for microdds_client.
2. Added new parameter `XRCE_DDS_AG_IP` that encodes the micro XRCE-DDS Agent IP. Its default value corresponds to `127.0.0.1`.
3. When the client is started without specifying the udp port (missing `-p` argument) then the parameter `XRCE_DDS_PRT` is used.
4. When the client is started without specifying the Agent IP (missing `-h` argument) then the parameter `XRCE_DDS_AG_IP` is used.
5. Increased verbosity of `microdds_client status`.
6. To make the conversion from IP addresses in decimal-dot notation to int32 and vice-versa `Tools/convert_ip.py` is added.

Fixes #21179

### About formatting the IP address as int32
Taking the IP address in decimal-dot notation, converting it in HEX dot notation and the stacking together the 4 bytes is not enough. The parameters in PX4 are **int32**, not **u**int32. Therefore, the resulting 4 byte variable must be written as "the decimal number whose 2-complement representation in 32 bit is my variable". This number can be negative.
Examples:
|IP | IP in hex dot notation | complete HEX number | associated `uint32` | `int32` to write in `XRCE_DDS_${i}_AG_IP`|
|---|---|---|---|---|
|`127.0.0.1` | `7f.00.00.01` | `0x7f000001` | `2130706433` | `2130706433`|
|`192.168.1.2` | `c0.a8.01.02` | `0xc0a80102` | `3232235778` | `-1062731518`|

#### Make the conversions with `Tools/convert_ip.py`
- From decimal-dot notation to int32
   ```sh
  python3 Tools/convert_ip.py xxx.xxx.xxx.xxx
   ```
- From int32 to decimal-dot notation
   ```sh
  python3 Tools/convert_ip.py -r xxxxxx
   ```
Screenshot from 2023-03-07 23-30-39

### Test coverage
- **Tested in SITL** starting the client with different combination of parameters. In SITL the agent responds to `127.0.0.1` and `192.168.1.3` (my local IP address, change it accordingly to the tester setup). During the test, the auto-startup of the client on the [startup script](https://github.com/PX4/PX4-Autopilot/blob/6e21e96b50cce8695b65c017ad0f67924b571cf8/ROMFS/px4fmu_common/init.d-posix/rcS#L267-L280) has been disabled
   1. Agent started with `MicroXRCEAgent udp4 -p 9999`. microdds_client started with udp port and IP passed as explicit parameters.
      ```
      microdds_client start -t udp -p 9999 -h 192.168.1.3
      ```      
      Screenshot from 2023-03-07 23-49-37

   1. Agent started with `MicroXRCEAgent udp4 -p 7777`. microdds_client started with no explicit udp port, `XRCE_DDS_PRT` is used.
      ```
      param set XRCE_DDS_PRT 7777
      microdds_client start -t udp -h 127.0.0.1
      ```
      Screenshot from 2023-03-07 23-52-09

   1. Agent started with `MicroXRCEAgent udp4 -p 8888`. microdds_client started with no explicit IP address, `XRCE_DDS_AG_IP` is used.
      ```
      param set XRCE_DDS_AG_IP -1062731517
      microdds_client start -t udp -p 8888
      ```
      Screenshot from 2023-03-07 23-54-06
      to replicate, use the matching IP on the tester machine.
   1. Agent started with `MicroXRCEAgent udp4 -p 7777`. microdds_client started with no explicit references to port or IP address, `XRCE_DDS_PRT` and `XRCE_DDS_AG_IP` are used. _This is the default command that autostarts the client when_ `XRCE_DDS_CFG=ETH`.
      ```
      param set XRCE_DDS_PRT 7777
      param set XRCE_DDS_AG_IP -1062731517
      microdds_client start -t udp
      ```
      Screenshot from 2023-03-07 23-55-53
      To replicate, use the matching IP on the tester machine.
 - **Test on real hardware**. I don't have a physical board therefore I cannot set `XRCE_DDS_CFG='ETH'` and test the auto-startup. The testing procedure can be the following:
    1. Configure the [Ethernet connection](https://docs.px4.io/main/en/advanced_config/ethernet_setup.html) of the board and test it.
    7. Set `XRCE_DDS_AG_IP` and `XRCE_DDS_PRT` to the desired values matching the agent configuration.
    8. Set `XRCE_DDS_CFG='ETH'` and reboot the board. It should automatically connect to the running agent. If the parameter is set using mavlink console then the the value must be `1000`.
 
   `XRCE_DDS_AG_IP` and `XRCE_DDS_PRT` may not show up in QGC as they initially result _unused_. In that case swap the order of point 2 and 3 and reboot at the end.

### CODE REVIEW

**Comment by mrpollo** [comment]
Tested on CUAV Pixhawk V6X running the client with both IP as string and Int32, everything works as expected

LGTM

Micro XRCE Agent:
* Ubuntu 20.04
* port on default 8888
* local IP: 192.168.2.206

Micro XRCE Client:
* PX4 (this branch)
* CUAV Pixhawk V6X
* Connected using ETH to a network router on the same LAN as Agent

## -h IP{string] format
MAVLink Console
```
microdds_client start -t udp -h 192.168.2.206
```

## -h -p:{int32} format
I needed to convert the IP from the Agent to Int32, ([I used the following converter](https://www.rapidtables.com/convert/number/hex-to-decimal.html))

```
192.168.2.206 = c0.a8.02.ce
0xc0a802ce
-1062731058
```

**Starting from Param:**
```
param set XRCE_DDS_0_AG_IP -1062731058
microdds_client start -t udp -h p:XRCE_DDS_0_AG_IP
nsh> INFO  [microdds_client] synchronized with time offset 1678132414665278us
INFO  [microdds_client] successfully created rt/fmu/out/failsafe_flags data writer, topic id: 84
INFO  [microdds_client] successfully created rt/fmu/out/sensor_combined data writer, topic id: 170
INFO  [microdds_client] successfully created rt/fmu/out/timesync_status data writer, topic id: 190
INFO  [microdds_client] successfully created rt/fmu/out/vehicle_attitude data writer, topic id: 207
INFO  [microdds_client] successfully created rt/fmu/out/vehicle_control_mode data writer, topic id: 213
INFO  [microdds_client] successfully created rt/fmu/out/vehicle_local_position data writer, topic id: 220
INFO  [microdds_client] successfully created rt/fmu/out/vehicle_odometry data writer, topic id: 225
INFO  [microdds_client] successfully created rt/fmu/out/vehicle_status data writer, topic id: 230
INFO  [microdds_client] No ping response, disconnecting
```
Screenshot 2023-03-06 at 12 00 33 PM
Screenshot 2023-03-06 at 12 00 25 PM

**Comment by beniaminopozzan** [comment]
@mrpollo thank you for testing it out and discovering that the parameters where not persistent!

I removed the initial multi instance support, and added a python script to make the conversion easier until IP / strings will be supported as parameters. Could you give another try to this version?

**Comment by mrpollo** [comment]
Here's my experience with the latest changes

# Testing Autostart

## 1. Flashing and booting

After flashing trough command line and booting, QGC displays only the `XRCE_DDS_CFG` param with the default value

Screenshot 2023-03-08 at 12 31 20 PM

## 2. Updating params

Using the parameter explorer in QGC I located the Micro XRCE-DDS section, updated the `XRCE_DDS_CFG` param to Ethernet, and was prompted to reboot, after boot the missing params are available `XRCE_DDS_AG_IP` & `XRCE_DDS_PORT`

Screenshot 2023-03-08 at 12 28 17 PM

I rebooted once again, after updating the IP param with the INT32 version (`-1062731058`) of the agent IP on my local network (`192.168.2.206`)

Screenshot 2023-03-08 at 12 36 14 PM

## 3. Auto-start

A few seconds after the reboot command was sent the local agent on my network starts receiving messages!


microdds agent screen recording


I also went over to the mavlink console to verify the process was reporting the right status

```
nsh: sysinit: fopen failed: No such file or directory

NuttShell (NSH) NuttX-11.0.0
nsh> microdds_client status
INFO  [microdds_client] Running, connected
INFO  [microdds_client] Using transport: udp
INFO  [microdds_client] Agent IP: 192.168.2.206
INFO  [microdds_client] Agent port: 8888
INFO  [microdds_client] Payload tx: 49538 B/s
INFO  [microdds_client] Payload rx: 0 B/s
nsh> 
```


@beniaminopozzan Please note the error when calling microdds_client

> nsh: sysinit: fopen failed: No such file or directory

# Testing manual start

## Flash and boot

After the last test, I reset all the params to default, and rebooted

```
nsh> microdds_client start -t udp -h 192.168.2.206
INFO  [microdds_client] synchronized with time offset 1678308094614176us
INFO  [microdds_client] successfully created rt/fmu/out/failsafe_flags data writer, topic id: 85
INFO  [microdds_client] successfully created rt/fmu/out/sensor_combined data writer, topic id: 172
INFO  [microdds_client] successfully created rt/fmu/out/timesync_status data writer, topic id: 194
INFO  [microdds_client] successfully created rt/fmu/out/vehicle_attitude data writer, topic id: 211
INFO  [microdds_client] successfully created rt/fmu/out/vehicle_control_mode data writer, topic id: 217
INFO  [microdds_client] successfully created rt/fmu/out/vehicle_local_position data writer, topic id: 224
INFO  [microdds_client] successfully created rt/fmu/out/vehicle_odometry data writer, topic id: 229
INFO  [microdds_client] successfully created rt/fmu/out/vehicle_status data writer, topic id: 234
```

After a few minutes I stopped the process and went over to the params view, I discovered that once you run the microdds_client manually the IP and PRT params show up, is this expected?

Screenshot 2023-03-08 at 12 46 23 PM

**Comment by beniaminopozzan** [comment]
Thanks @mrpollo ! Glad that it works (mostly)

Regarding

> Please note the error when calling microdds_client
> > nsh: sysinit: fopen failed: No such file or directory

I don't know, it's the first time that I see that error; and it does not appear in SITL. But, is it related to this PR?

> After a few minutes I stopped the process and went over to the params view, I discovered that once you run the microdds_client manually the IP and PRT params show up, is this expected?

Unfortunately, yes. They are marked as _used_ as soon as the module is started, even if `-h` or `-p` are specified. However, I can fix that!

**Comment by beniaminopozzan** [comment]
@mrpollo , I checked and the error
```sh
sysinit: fopen failed: No such file or directory
```
is present on the current `main`, therefore I exclude that it is caused by this PR.

Everything else is working as expected. Manual start, automatic start from SD, automatic start from parameters.

---

PR for the doc update: PX4/PX4-user_guide#2352

---

I think it is ready for review. @dagar could you take a look at this?

**Comment by mrpollo** [comment]
Hey @dagar we are seeing flash overflow on 6C, never seen this one before. Can you advise how to proceed?

**Comment by dagar** [comment]
> Hey @dagar we are seeing flash overflow on 6C, never seen this one before. Can you advise how to proceed?

There should be plenty of space after a rebase.

**Comment by beniaminopozzan** [comment]
The PR is already on top of the current main, does ti mean it is too "big"?

**Comment by mrpollo** [comment]
It just means that a few things from the last time you rebased main, one last rebase should fix it

**Comment by beniaminopozzan** [comment]
It's fixed now!

**Comment by hamishwillee** [comment]
@mrpollo Review on this would be good!

File: Tools/convert_ip.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    print(ip)
+
+else:
+    try:
+        ip = int(args.input)
+    except:
+        raise ValueError("Not a valid IP")
+    if(ip < 0):
+        ip = ip + 0x100000000
+    print("the decimal dot conversion of", args.input, "is")
```
Comment (mrpollo) [inline Tools/convert_ip.py:52]: would it be possible to remove this line? it makes it hard to grep the output of the script

**Comment by mrpollo** [comment]
Thanks for the effort @beniaminopozzan the feature works as advertised and I see no major issues, I did end up sending a commit to update the python script (to remove verbose output) which triggered a new CI pass (which I'm ignoring since it's not testing my change)

Thank you 🎉

**Comment by hamishwillee** [comment]
Very cool. Will this be in PX4 v1.14?

**Comment by beniaminopozzan** [comment]
yes!

**Comment by ozitoune** [comment]
Thanks a lot !! I will try tomorrow with my v6x plugged directly on the Ethernet of raspberry, to see if topics are all received by ros2 without anything to do, with a cold start the pixhawk and the raspberry powered by battery :) !!
I have some bizarre issue with a routeur between the two , I will try to reproduce and post it
Thanks again for your precious help and your work !!

**Comment by beniaminopozzan** [comment]
> Thanks a lot !! I will try tomorrow with my v6x plugged directly on the Ethernet of raspberry, to see if topics are all received by ros2 without anything to do, with a cold start the pixhawk and the raspberry powered by battery :) !! I have some bizarre issue with a routeur between the two , I will try to reproduce and post it Thanks again for your precious help and your work !!

**without anything to do** ... @ozitoune , don't forget to set IP and port! :))

**Comment by ozitoune** [comment]
Hello, yes i have done all and it work like a charm !!!!  Thank you !! 

I have updated the dds_topics.yalm, to have lot off more topics avaible by ros. It compile perfectly and work well . 
Should I make a pull request ?

**Pull request merged**